In [1]:
import pandas as pd
from decimal import Decimal
from io import StringIO

from carbon import CarbonSimulatorUI, __version__, __date__
print(f"Carbon Version v{__version__} ({__date__})", )

Carbon Version v2.2-BETA2 (03/Jan/2022)


# FastRouter (NBTest 47)

NBTEST: NOTEST_DEFAULT = TEST

## FastRouterTests

Initialize a fast simulator

In [2]:
Sim = CarbonSimulatorUI(pair="ETH/USDC", verbose=False, raiseonerror=True)
FastSim = CarbonSimulatorUI(pair="ETH/USDC", verbose=False, matching_method='fast', raiseonerror=True)

Here we restrain the orders to just the top ones such that we can test the exact algo alongside

In [3]:
ORDERS_CSV = StringIO("""
id,liquidity,lowest_rate,highest_rate,current_rate
0,254814732,256,257,257
1,253827078,255,256,256
2,252839424,254,255,255
3,251851770,253,254,254
4,6913578,5,6,6
5,7901232,6,7,7
6,248888808,250,251,251
7,9876540,8,9,9
8,10864194,9,10,10
9,11851848,10,11,11
10,244938192,246,247,247
11,13827156,12,13,13
12,14814810,13,14,14
13,15802464,14,15,15
14,16790118,15,16,16
15,239999922,241,242,242
16,18765426,17,18,18
17,19753080,18,19,19
18,20740734,19,20,20
19,21728388,20,21,21
20,22716042,21,22,22
21,234073998,235,236,236
22,24691350,23,24,24
23,25679004,24,25,25
24,231111036,232,233,233
25,27654312,26,27,27
26,28641966,27,28,28
27,29629620,28,29,29
28,227160420,228,229,229
29,31604928,30,31,31
30,32592582,31,32,32
31,33580236,32,33,33
32,34567890,33,34,34
33,35555544,34,35,35
34,36543198,35,36,36
35,37530852,36,37,37
36,219259188,220,221,221
37,39506160,38,39,39
38,40493814,39,40,40
39,41481468,40,41,41
40,42469122,41,42,42
41,43456776,42,43,43
42,44444430,43,44,44
43,45432084,44,45,45
44,46419738,45,46,46
45,47407392,46,47,47
46,48395046,47,48,48
47,49382700,48,49,49
48,50370354,49,50,50
49,51358008,50,51,51
50,52345662,51,52,52
51,53333316,52,53,53
52,54320970,53,54,54
53,55308624,54,55,55
54,56296278,55,56,56
55,200493762,201,202,202
56,58271586,57,58,58
57,59259240,58,59,59
58,60246894,59,60,60
59,61234548,60,61,61
60,62222202,61,62,62
61,63209856,62,63,63
62,64197510,63,64,64
63,65185164,64,65,65
64,66172818,65,66,66
65,190617222,191,192,192
66,189629568,190,191,191
67,69135780,68,69,69
68,70123434,69,70,70
69,71111088,70,71,71
70,72098742,71,72,72
71,73086396,72,73,73
72,74074050,73,74,74
73,75061704,74,75,75
74,76049358,75,76,76
75,77037012,76,77,77
76,78024666,77,78,78
77,79012320,78,79,79
78,177777720,178,179,179
79,80987628,80,81,81
80,81975282,81,82,82
81,82962936,82,83,83
82,83950590,83,84,84
83,84938244,84,85,85
84,171851796,172,173,173
85,86913552,86,87,87
86,87901206,87,88,88
87,88888860,88,89,89
88,89876514,89,90,90
89,90864168,90,91,91
90,91851822,91,92,92
91,164938218,165,166,166
92,93827130,93,94,94
93,94814784,94,95,95
94,95802438,95,96,96
95,96790092,96,97,97
96,97777746,97,98,98
97,98765400,98,99,99
98,99753054,99,100,100
99,100740708,100,101,101
100,101728362,101,102,102
101,102716016,102,103,103
102,154074024,154,155,155
103,104691324,104,105,105
104,105678978,105,106,106
105,151111062,151,152,152
106,107654286,107,108,108
107,108641940,108,109,109
108,109629594,109,110,110
109,110617248,110,111,111
110,111604902,111,112,112
111,112592556,112,113,113
112,113580210,113,114,114
113,114567864,114,115,115
114,115555518,115,116,116
115,116543172,116,117,117
116,117530826,117,118,118
117,118518480,118,119,119
118,119506134,119,120,120
119,137283906,137,138,138
120,136296252,136,137,137
121,122469096,122,123,123
122,123456750,123,124,124
123,124444404,124,125,125
124,125432058,125,126,126
125,126419712,126,127,127
126,127407366,127,128,128
127,128395020,128,129,129
128,129382674,129,130,130
129,130370328,130,131,131
130,131357982,131,132,132
131,132345636,132,133,133
132,133333290,133,134,134
133,134320944,134,135,135
134,135308598,135,136,136
135,121481442,121,122,122
136,120493788,120,121,121
137,138271560,138,139,139
138,139259214,139,140,140
139,140246868,140,141,141
140,141234522,141,142,142
141,142222176,142,143,143
142,143209830,143,144,144
143,144197484,144,145,145
144,145185138,145,146,146
145,146172792,146,147,147
146,147160446,147,148,148
147,148148100,148,149,149
148,149135754,149,150,150
149,150123408,150,151,151
150,106666632,106,107,107
151,152098716,152,153,153
152,153086370,153,154,154
153,103703670,103,104,104
154,155061678,155,156,156
155,156049332,156,157,157
156,157036986,157,158,158
157,158024640,158,159,159
158,159012294,159,160,160
159,159999948,160,161,161
160,160987602,161,162,162
161,161975256,162,163,163
162,162962910,163,164,164
163,163950564,164,165,165
164,92839476,92,93,93
165,165925872,166,167,167
166,166913526,167,168,168
167,167901180,168,169,169
168,168888834,169,170,170
169,169876488,170,171,171
170,170864142,171,172,172
171,85925898,85,86,86
172,172839450,173,174,174
173,173827104,174,175,175
174,174814758,175,176,176
175,175802412,176,177,177
176,176790066,177,178,178
177,79999974,79,80,80
178,178765374,179,180,180
179,179753028,180,181,181
180,180740682,181,182,182
181,181728336,182,183,183
182,182715990,183,184,184
183,183703644,184,185,185
184,184691298,185,186,186
185,185678952,186,187,187
186,186666606,187,188,188
187,187654260,188,189,189
188,188641914,189,190,190
189,68148126,67,68,68
190,67160472,66,67,67
191,191604876,192,193,193
192,192592530,193,194,194
193,193580184,194,195,195
194,194567838,195,196,196
195,195555492,196,197,197
196,196543146,197,198,198
197,197530800,198,199,199
198,198518454,199,200,200
199,199506108,200,201,201
200,57283932,56,57,57
201,201481416,202,203,203
202,202469070,203,204,204
203,203456724,204,205,205
204,204444378,205,206,206
205,205432032,206,207,207
206,206419686,207,208,208
207,207407340,208,209,209
208,208394994,209,210,210
209,209382648,210,211,211
210,210370302,211,212,212
211,211357956,212,213,213
212,212345610,213,214,214
213,213333264,214,215,215
214,214320918,215,216,216
215,215308572,216,217,217
216,216296226,217,218,218
217,217283880,218,219,219
218,218271534,219,220,220
219,38518506,37,38,38
220,220246842,221,222,222
221,221234496,222,223,223
222,222222150,223,224,224
223,223209804,224,225,225
224,224197458,225,226,226
225,225185112,226,227,227
226,226172766,227,228,228
227,30617274,29,30,30
228,228148074,229,230,230
229,229135728,230,231,231
230,230123382,231,232,232
231,26666658,25,26,26
232,232098690,233,234,234
233,233086344,234,235,235
234,23703696,22,23,23
235,235061652,236,237,237
236,236049306,237,238,238
237,237036960,238,239,239
238,238024614,239,240,240
239,239012268,240,241,241
240,17777772,16,17,17
241,240987576,242,243,243
242,241975230,243,244,244
243,242962884,244,245,245
244,243950538,245,246,246
245,12839502,11,12,12
246,245925846,247,248,248
247,246913500,248,249,249
248,247901154,249,250,250
249,8888886,7,8,8
250,249876462,251,252,252
251,250864116,252,253,253
252,5925924,4,5,5
253,4938270,3,4,4
254,3950616,2,3,3
255,2962962,1,2,2
""")

In [4]:
#test_orders = pd.read_csv('orders.csv', dtype=str)
test_orders = pd.read_csv(ORDERS_CSV, dtype=str)
test_orders = test_orders[test_orders.id.isin(['0','1','2','3','251','250','6','248','247','246','10'])].copy()
test_orders

id  liquidity lowest_rate highest_rate current_rate
0      0  254814732         256          257          257
1      1  253827078         255          256          256
2      2  252839424         254          255          255
3      3  251851770         253          254          254
6      6  248888808         250          251          251
10    10  244938192         246          247          247
246  246  245925846         247          248          248
247  247  246913500         248          249          249
248  248  247901154         249          250          250
250  250  249876462         251          252          252
251  251  250864116         252          253          253

In [5]:
for i in test_orders.index:
    Sim.add_strategy('USDC', Decimal(test_orders.liquidity[i]), Decimal(test_orders.highest_rate[i]), Decimal(test_orders.lowest_rate[i]),  None, None, None)
    FastSim.add_strategy('USDC', Decimal(test_orders.liquidity[i]), Decimal(test_orders.highest_rate[i]), Decimal(test_orders.lowest_rate[i]),  None, None, None)

FastSim.state()['orders']

id     pair   tkn        y_int            y y_unit  disabled p_start  \
0    0  ETHUSDC  USDC  254814732.0  254814732.0   USDC     False   257.0   
1    1  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
2    2  ETHUSDC  USDC  253827078.0  253827078.0   USDC     False   256.0   
3    3  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
4    4  ETHUSDC  USDC  252839424.0  252839424.0   USDC     False   255.0   
5    5  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
6    6  ETHUSDC  USDC  251851770.0  251851770.0   USDC     False   254.0   
7    7  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
8    8  ETHUSDC  USDC  248888808.0  248888808.0   USDC     False   251.0   
9    9  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
10  10  ETHUSDC  USDC  244938192.0  244938192.0   USDC     False   247.0   
11  11  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
12  12  ETHUSDC  USDC  245925846.0  245925846.0   USDC     False   248.0   
13  13  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
14  14  ETHUSDC  USDC  246913500.0  246913500.0   USDC     False   249.0   
15  15  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
16  16  ETHUSDC  USDC  247901154.0  247901154.0   USDC     False   250.0   
17  17  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
18  18  ETHUSDC  USDC  249876462.0  249876462.0   USDC     False   252.0   
19  19  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
20  20  ETHUSDC  USDC  250864116.0  250864116.0   USDC     False   253.0   
21  21  ETHUSDC   ETH          0.0          0.0    ETH      True    None   

    p_end p_marg        p_unit  lid  
0   256.0  257.0  USDC per ETH    1  
1    None   None  USDC per ETH    0  
2   255.0  256.0  USDC per ETH    3  
3    None   None  USDC per ETH    2  
4   254.0  255.0  USDC per ETH    5  
5    None   None  USDC per ETH    4  
6   253.0  254.0  USDC per ETH    7  
7    None   None  USDC per ETH    6  
8   250.0  251.0  USDC per ETH    9  
9    None   None  USDC per ETH    8  
10  246.0  247.0  USDC per ETH   11  
11   None   None  USDC per ETH   10  
12  247.0  248.0  USDC per ETH   13  
13   None   None  USDC per ETH   12  
14  248.0  249.0  USDC per ETH   15  
15   None   None  USDC per ETH   14  
16  249.0  250.0  USDC per ETH   17  
17   None   None  USDC per ETH   16  
18  251.0  252.0  USDC per ETH   19  
19   None   None  USDC per ETH   18  
20  252.0  253.0  USDC per ETH   21  
21   None   None  USDC per ETH   20

In [6]:
FastSim.state()['orders']

id     pair   tkn        y_int            y y_unit  disabled p_start  \
0    0  ETHUSDC  USDC  254814732.0  254814732.0   USDC     False   257.0   
1    1  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
2    2  ETHUSDC  USDC  253827078.0  253827078.0   USDC     False   256.0   
3    3  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
4    4  ETHUSDC  USDC  252839424.0  252839424.0   USDC     False   255.0   
5    5  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
6    6  ETHUSDC  USDC  251851770.0  251851770.0   USDC     False   254.0   
7    7  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
8    8  ETHUSDC  USDC  248888808.0  248888808.0   USDC     False   251.0   
9    9  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
10  10  ETHUSDC  USDC  244938192.0  244938192.0   USDC     False   247.0   
11  11  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
12  12  ETHUSDC  USDC  245925846.0  245925846.0   USDC     False   248.0   
13  13  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
14  14  ETHUSDC  USDC  246913500.0  246913500.0   USDC     False   249.0   
15  15  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
16  16  ETHUSDC  USDC  247901154.0  247901154.0   USDC     False   250.0   
17  17  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
18  18  ETHUSDC  USDC  249876462.0  249876462.0   USDC     False   252.0   
19  19  ETHUSDC   ETH          0.0          0.0    ETH      True    None   
20  20  ETHUSDC  USDC  250864116.0  250864116.0   USDC     False   253.0   
21  21  ETHUSDC   ETH          0.0          0.0    ETH      True    None   

    p_end p_marg        p_unit  lid  
0   256.0  257.0  USDC per ETH    1  
1    None   None  USDC per ETH    0  
2   255.0  256.0  USDC per ETH    3  
3    None   None  USDC per ETH    2  
4   254.0  255.0  USDC per ETH    5  
5    None   None  USDC per ETH    4  
6   253.0  254.0  USDC per ETH    7  
7    None   None  USDC per ETH    6  
8   250.0  251.0  USDC per ETH    9  
9    None   None  USDC per ETH    8  
10  246.0  247.0  USDC per ETH   11  
11   None   None  USDC per ETH   10  
12  247.0  248.0  USDC per ETH   13  
13   None   None  USDC per ETH   12  
14  248.0  249.0  USDC per ETH   15  
15   None   None  USDC per ETH   14  
16  249.0  250.0  USDC per ETH   17  
17   None   None  USDC per ETH   16  
18  251.0  252.0  USDC per ETH   19  
19   None   None  USDC per ETH   18  
20  252.0  253.0  USDC per ETH   21  
21   None   None  USDC per ETH   20

### Route by Source

#### AMM buys ETH

Lets do some simple checks against the exact algo and then verify against the fast router results

1. Trade an amount against exact

In [7]:
results = Sim.amm_buys('ETH',10000000, execute=False)['trades']  # route_trade_by_source
assert(f"{results[results.uid=='0']['amt1'].values[0]:.0f}" == "2519667817")
results

uid  id subid                                       note   aggr   exec  \
0   0.0   0     0                                   route #0  False  False   
0   0.1   0     1                                   route #2  False  False   
0   0.2   0     2                                   route #4  False  False   
0   0.3   0     3                                   route #6  False  False   
0   0.4   0     4                                   route #8  False  False   
0   0.5   0     5                                  route #18  False  False   
0   0.6   0     6                                  route #20  False  False   
0   0.7   0     7                                  route #14  False  False   
0   0.8   0     8                                  route #16  False  False   
0   0.9   0     9                                  route #12  False  False   
0  0.10   0    10                                  route #10  False  False   
0     0   0     A  AMM sells 2519667817USDC buys 10000000ETH   True  False   

  limitfail          amt1  tkn1          amt2 tkn2     pair  \
0      None  2.548147e+08  USDC  9.934316e+05  ETH  ETHUSDC   
0      None  2.538271e+08  USDC  9.934543e+05  ETH  ETHUSDC   
0      None  2.528394e+08  USDC  9.934771e+05  ETH  ETHUSDC   
0      None  2.518518e+08  USDC  9.935000e+05  ETH  ETHUSDC   
0      None  2.488888e+08  USDC  9.935701e+05  ETH  ETHUSDC   
0      None  2.498765e+08  USDC  9.935465e+05  ETH  ETHUSDC   
0      None  2.508641e+08  USDC  9.935232e+05  ETH  ETHUSDC   
0      None  2.469135e+08  USDC  9.936177e+05  ETH  ETHUSDC   
0      None  2.479012e+08  USDC  9.935938e+05  ETH  ETHUSDC   
0      None  2.459258e+08  USDC  9.936418e+05  ETH  ETHUSDC   
0      None  1.596493e+07  USDC  6.464387e+04  ETH  ETHUSDC   
0      None  2.519668e+09  USDC  1.000000e+07  ETH  ETHUSDC   

                                   routeix  nroutes       price        p_unit  
0                                        0        1  251.966782  USDC per ETH  
0                                        2        1  251.966782  USDC per ETH  
0                                        4        1  251.966782  USDC per ETH  
0                                        6        1  251.966782  USDC per ETH  
0                                        8        1  251.966782  USDC per ETH  
0                                       18        1  251.966782  USDC per ETH  
0                                       20        1  251.966782  USDC per ETH  
0                                       14        1  251.966782  USDC per ETH  
0                                       16        1  251.966782  USDC per ETH  
0                                       12        1  251.966782  USDC per ETH  
0                                       10        1  251.966782  USDC per ETH  
0  [0, 2, 4, 6, 8, 18, 20, 14, 16, 12, 10]       11  251.966782  USDC per ETH

We can observe the price is 251.996 USDC per ETH.

For trade_by_source, when we increase the trade amount we push further down the curve which means that effective price should get worse.

In [8]:
results = Sim.amm_buys('ETH',10500000, execute=False)['trades']  # route_trade_by_source
assert(f"{results[results.uid=='1']['amt1'].values[0]:.0f}" == "2643009195")
results

uid  id subid                                       note   aggr   exec  \
0   1.0   1     0                                   route #0  False  False   
0   1.1   1     1                                   route #2  False  False   
0   1.2   1     2                                   route #4  False  False   
0   1.3   1     3                                   route #6  False  False   
0   1.4   1     4                                   route #8  False  False   
0   1.5   1     5                                  route #18  False  False   
0   1.6   1     6                                  route #20  False  False   
0   1.7   1     7                                  route #14  False  False   
0   1.8   1     8                                  route #16  False  False   
0   1.9   1     9                                  route #12  False  False   
0  1.10   1    10                                  route #10  False  False   
0     1   1     A  AMM sells 2643009195USDC buys 10500000ETH   True  False   

  limitfail          amt1  tkn1          amt2 tkn2     pair  \
0      None  2.548147e+08  USDC  9.934316e+05  ETH  ETHUSDC   
0      None  2.538271e+08  USDC  9.934543e+05  ETH  ETHUSDC   
0      None  2.528394e+08  USDC  9.934771e+05  ETH  ETHUSDC   
0      None  2.518518e+08  USDC  9.935000e+05  ETH  ETHUSDC   
0      None  2.488888e+08  USDC  9.935701e+05  ETH  ETHUSDC   
0      None  2.498765e+08  USDC  9.935465e+05  ETH  ETHUSDC   
0      None  2.508641e+08  USDC  9.935232e+05  ETH  ETHUSDC   
0      None  2.469135e+08  USDC  9.936177e+05  ETH  ETHUSDC   
0      None  2.479012e+08  USDC  9.935938e+05  ETH  ETHUSDC   
0      None  2.459258e+08  USDC  9.936418e+05  ETH  ETHUSDC   
0      None  1.393063e+08  USDC  5.646439e+05  ETH  ETHUSDC   
0      None  2.643009e+09  USDC  1.050000e+07  ETH  ETHUSDC   

                                   routeix  nroutes       price        p_unit  
0                                        0        1  251.715161  USDC per ETH  
0                                        2        1  251.715161  USDC per ETH  
0                                        4        1  251.715161  USDC per ETH  
0                                        6        1  251.715161  USDC per ETH  
0                                        8        1  251.715161  USDC per ETH  
0                                       18        1  251.715161  USDC per ETH  
0                                       20        1  251.715161  USDC per ETH  
0                                       14        1  251.715161  USDC per ETH  
0                                       16        1  251.715161  USDC per ETH  
0                                       12        1  251.715161  USDC per ETH  
0                                       10        1  251.715161  USDC per ETH  
0  [0, 2, 4, 6, 8, 18, 20, 14, 16, 12, 10]       11  251.715161  USDC per ETH

And indeed for the exact algo, the price gets lower (251.996 vs 251.715), i.e. for every unit of ETH spent I get less units of USDC in return

We can then do a similar thing for the fast router

In [9]:
results = FastSim.amm_buys('ETH',10000000, execute=False)['trades']  # route_trade_by_source
assert(f"{results[results.uid=='0']['amt1'].values[0]:.0f}" == "2519667817")
results

uid  id subid                                       note   aggr   exec  \
0   0.0   0     0                                   route #0  False  False   
0   0.1   0     1                                   route #2  False  False   
0   0.2   0     2                                   route #4  False  False   
0   0.3   0     3                                   route #6  False  False   
0   0.4   0     4                                  route #20  False  False   
0   0.5   0     5                                  route #18  False  False   
0   0.6   0     6                                   route #8  False  False   
0   0.7   0     7                                  route #16  False  False   
0   0.8   0     8                                  route #14  False  False   
0   0.9   0     9                                  route #12  False  False   
0  0.10   0    10                                  route #10  False  False   
0     0   0     A  AMM sells 2519667817USDC buys 10000000ETH   True  False   

  limitfail          amt1  tkn1          amt2 tkn2     pair  \
0      None  2.548147e+08  USDC  9.934316e+05  ETH  ETHUSDC   
0      None  2.538271e+08  USDC  9.934543e+05  ETH  ETHUSDC   
0      None  2.528394e+08  USDC  9.934771e+05  ETH  ETHUSDC   
0      None  2.518518e+08  USDC  9.935000e+05  ETH  ETHUSDC   
0      None  2.508641e+08  USDC  9.935232e+05  ETH  ETHUSDC   
0      None  2.498765e+08  USDC  9.935465e+05  ETH  ETHUSDC   
0      None  2.488888e+08  USDC  9.935701e+05  ETH  ETHUSDC   
0      None  2.479012e+08  USDC  9.935938e+05  ETH  ETHUSDC   
0      None  2.469135e+08  USDC  9.936177e+05  ETH  ETHUSDC   
0      None  2.459258e+08  USDC  9.936418e+05  ETH  ETHUSDC   
0      None  1.596493e+07  USDC  6.464387e+04  ETH  ETHUSDC   
0      None  2.519668e+09  USDC  1.000000e+07  ETH  ETHUSDC   

                                   routeix  nroutes       price        p_unit  
0                                        0        1  251.966782  USDC per ETH  
0                                        2        1  251.966782  USDC per ETH  
0                                        4        1  251.966782  USDC per ETH  
0                                        6        1  251.966782  USDC per ETH  
0                                       20        1  251.966782  USDC per ETH  
0                                       18        1  251.966782  USDC per ETH  
0                                        8        1  251.966782  USDC per ETH  
0                                       16        1  251.966782  USDC per ETH  
0                                       14        1  251.966782  USDC per ETH  
0                                       12        1  251.966782  USDC per ETH  
0                                       10        1  251.966782  USDC per ETH  
0  [0, 2, 4, 6, 20, 18, 8, 16, 14, 12, 10]       11  251.966782  USDC per ETH

We see that the orders are filled the same and the corresponding price is as we saw before.

We can then trade the higher amount and again, expect to see the effectice rate decrease.

In [10]:
results = FastSim.amm_buys('ETH',10500000, execute=False)['trades']  # route_trade_by_source
assert(f"{results[results.uid=='1']['amt1'].values[0]:.0f}" == "2643009195")
results

uid  id subid                                       note   aggr   exec  \
0   1.0   1     0                                   route #0  False  False   
0   1.1   1     1                                   route #2  False  False   
0   1.2   1     2                                   route #4  False  False   
0   1.3   1     3                                   route #6  False  False   
0   1.4   1     4                                  route #20  False  False   
0   1.5   1     5                                  route #18  False  False   
0   1.6   1     6                                   route #8  False  False   
0   1.7   1     7                                  route #16  False  False   
0   1.8   1     8                                  route #14  False  False   
0   1.9   1     9                                  route #12  False  False   
0  1.10   1    10                                  route #10  False  False   
0     1   1     A  AMM sells 2643009195USDC buys 10500000ETH   True  False   

  limitfail          amt1  tkn1          amt2 tkn2     pair  \
0      None  2.548147e+08  USDC  9.934316e+05  ETH  ETHUSDC   
0      None  2.538271e+08  USDC  9.934543e+05  ETH  ETHUSDC   
0      None  2.528394e+08  USDC  9.934771e+05  ETH  ETHUSDC   
0      None  2.518518e+08  USDC  9.935000e+05  ETH  ETHUSDC   
0      None  2.508641e+08  USDC  9.935232e+05  ETH  ETHUSDC   
0      None  2.498765e+08  USDC  9.935465e+05  ETH  ETHUSDC   
0      None  2.488888e+08  USDC  9.935701e+05  ETH  ETHUSDC   
0      None  2.479012e+08  USDC  9.935938e+05  ETH  ETHUSDC   
0      None  2.469135e+08  USDC  9.936177e+05  ETH  ETHUSDC   
0      None  2.459258e+08  USDC  9.936418e+05  ETH  ETHUSDC   
0      None  1.393063e+08  USDC  5.646439e+05  ETH  ETHUSDC   
0      None  2.643009e+09  USDC  1.050000e+07  ETH  ETHUSDC   

                                   routeix  nroutes       price        p_unit  
0                                        0        1  251.715161  USDC per ETH  
0                                        2        1  251.715161  USDC per ETH  
0                                        4        1  251.715161  USDC per ETH  
0                                        6        1  251.715161  USDC per ETH  
0                                       20        1  251.715161  USDC per ETH  
0                                       18        1  251.715161  USDC per ETH  
0                                        8        1  251.715161  USDC per ETH  
0                                       16        1  251.715161  USDC per ETH  
0                                       14        1  251.715161  USDC per ETH  
0                                       12        1  251.715161  USDC per ETH  
0                                       10        1  251.715161  USDC per ETH  
0  [0, 2, 4, 6, 20, 18, 8, 16, 14, 12, 10]       11  251.715161  USDC per ETH

Again, we see that the price has, as before, decreased appropriately.

### Route by Target

#### AMM sells USDC

We can now test the route by target.

First we do the counter trade - selling the output amount of USDC from the initial ETH trade

In [11]:
results = Sim.amm_sells('USDC',2519667817, execute=False)['trades']  # route_trade_by_target
assert(f"{results[results.uid=='2']['amt2'].values[0]:.0f}" == "10000000")
results

uid  id subid                                       note   aggr   exec  \
0   2.0   2     0                                   route #0  False  False   
0   2.1   2     1                                   route #2  False  False   
0   2.2   2     2                                   route #4  False  False   
0   2.3   2     3                                   route #6  False  False   
0   2.4   2     4                                  route #18  False  False   
0   2.5   2     5                                  route #20  False  False   
0   2.6   2     6                                   route #8  False  False   
0   2.7   2     7                                  route #16  False  False   
0   2.8   2     8                                  route #14  False  False   
0   2.9   2     9                                  route #12  False  False   
0  2.10   2    10                                  route #10  False  False   
0     2   2     A  AMM sells 2519667817USDC buys 10000000ETH   True  False   

  limitfail          amt1  tkn1          amt2 tkn2     pair  \
0      None  2.548147e+08  USDC  9.934316e+05  ETH  ETHUSDC   
0      None  2.538271e+08  USDC  9.934543e+05  ETH  ETHUSDC   
0      None  2.528394e+08  USDC  9.934771e+05  ETH  ETHUSDC   
0      None  2.518518e+08  USDC  9.935000e+05  ETH  ETHUSDC   
0      None  2.498765e+08  USDC  9.935465e+05  ETH  ETHUSDC   
0      None  2.508641e+08  USDC  9.935232e+05  ETH  ETHUSDC   
0      None  2.488888e+08  USDC  9.935701e+05  ETH  ETHUSDC   
0      None  2.479012e+08  USDC  9.935938e+05  ETH  ETHUSDC   
0      None  2.469135e+08  USDC  9.936177e+05  ETH  ETHUSDC   
0      None  2.459258e+08  USDC  9.936418e+05  ETH  ETHUSDC   
0      None  1.596493e+07  USDC  6.464387e+04  ETH  ETHUSDC   
0      None  2.519668e+09  USDC  1.000000e+07  ETH  ETHUSDC   

                                   routeix  nroutes       price        p_unit  
0                                        0        1  251.966782  USDC per ETH  
0                                        2        1  251.966782  USDC per ETH  
0                                        4        1  251.966782  USDC per ETH  
0                                        6        1  251.966782  USDC per ETH  
0                                       18        1  251.966782  USDC per ETH  
0                                       20        1  251.966782  USDC per ETH  
0                                        8        1  251.966782  USDC per ETH  
0                                       16        1  251.966782  USDC per ETH  
0                                       14        1  251.966782  USDC per ETH  
0                                       12        1  251.966782  USDC per ETH  
0                                       10        1  251.966782  USDC per ETH  
0  [0, 2, 4, 6, 18, 20, 8, 16, 14, 12, 10]       11  251.966782  USDC per ETH

And we see that the return amount is 10000000 we saw before - but importantly so is the price

Now we expect that if we take less off the curve then the rate gets better

In [12]:
results = Sim.amm_sells('USDC',1500000000, execute=False)['trades']  # route_trade_by_target
assert(f"{results[results.uid=='3']['amt2'].values[0]:.0f}" == "5904869")
results

uid  id subid                                      note   aggr   exec  \
0  3.0   3     0                                  route #0  False  False   
0  3.1   3     1                                  route #2  False  False   
0  3.2   3     2                                  route #4  False  False   
0  3.3   3     3                                  route #6  False  False   
0  3.4   3     4                                 route #20  False  False   
0  3.5   3     5                                 route #18  False  False   
0    3   3     A  AMM sells 1500000000USDC buys 5904869ETH   True  False   

  limitfail          amt1  tkn1          amt2 tkn2     pair  \
0      None  2.548147e+08  USDC  9.934316e+05  ETH  ETHUSDC   
0      None  2.538271e+08  USDC  9.934543e+05  ETH  ETHUSDC   
0      None  2.528394e+08  USDC  9.934771e+05  ETH  ETHUSDC   
0      None  2.518518e+08  USDC  9.935000e+05  ETH  ETHUSDC   
0      None  2.508641e+08  USDC  9.935232e+05  ETH  ETHUSDC   
0      None  2.358029e+08  USDC  9.374828e+05  ETH  ETHUSDC   
0      None  1.500000e+09  USDC  5.904869e+06  ETH  ETHUSDC   

                routeix  nroutes       price        p_unit  
0                     0        1  254.027651  USDC per ETH  
0                     2        1  254.027651  USDC per ETH  
0                     4        1  254.027651  USDC per ETH  
0                     6        1  254.027651  USDC per ETH  
0                    20        1  254.027651  USDC per ETH  
0                    18        1  254.027651  USDC per ETH  
0  [0, 2, 4, 6, 20, 18]        6  254.027651  USDC per ETH

Indeed we see that the price is better for a smaller amount is less (254.027 vs 251.996)

Now we do the same for the Fast router. Trade in the previous output amount

In [13]:
results = FastSim.amm_sells('USDC',2519667817, execute=False)['trades']  # route_trade_by_target
assert(f"{results[results.uid=='2']['amt2'].values[0]:.0f}" == "10000000")
results

uid  id subid                                       note   aggr   exec  \
0   2.0   2     0                                   route #0  False  False   
0   2.1   2     1                                   route #2  False  False   
0   2.2   2     2                                   route #4  False  False   
0   2.3   2     3                                   route #6  False  False   
0   2.4   2     4                                  route #20  False  False   
0   2.5   2     5                                  route #18  False  False   
0   2.6   2     6                                   route #8  False  False   
0   2.7   2     7                                  route #16  False  False   
0   2.8   2     8                                  route #14  False  False   
0   2.9   2     9                                  route #12  False  False   
0  2.10   2    10                                  route #10  False  False   
0     2   2     A  AMM sells 2519667817USDC buys 10000000ETH   True  False   

  limitfail          amt1  tkn1          amt2 tkn2     pair  \
0      None  2.548147e+08  USDC  9.934316e+05  ETH  ETHUSDC   
0      None  2.538271e+08  USDC  9.934543e+05  ETH  ETHUSDC   
0      None  2.528394e+08  USDC  9.934771e+05  ETH  ETHUSDC   
0      None  2.518518e+08  USDC  9.935000e+05  ETH  ETHUSDC   
0      None  2.508641e+08  USDC  9.935232e+05  ETH  ETHUSDC   
0      None  2.498765e+08  USDC  9.935465e+05  ETH  ETHUSDC   
0      None  2.488888e+08  USDC  9.935701e+05  ETH  ETHUSDC   
0      None  2.479012e+08  USDC  9.935938e+05  ETH  ETHUSDC   
0      None  2.469135e+08  USDC  9.936177e+05  ETH  ETHUSDC   
0      None  2.459258e+08  USDC  9.936418e+05  ETH  ETHUSDC   
0      None  1.596493e+07  USDC  6.464387e+04  ETH  ETHUSDC   
0      None  2.519668e+09  USDC  1.000000e+07  ETH  ETHUSDC   

                                   routeix  nroutes       price        p_unit  
0                                        0        1  251.966782  USDC per ETH  
0                                        2        1  251.966782  USDC per ETH  
0                                        4        1  251.966782  USDC per ETH  
0                                        6        1  251.966782  USDC per ETH  
0                                       20        1  251.966782  USDC per ETH  
0                                       18        1  251.966782  USDC per ETH  
0                                        8        1  251.966782  USDC per ETH  
0                                       16        1  251.966782  USDC per ETH  
0                                       14        1  251.966782  USDC per ETH  
0                                       12        1  251.966782  USDC per ETH  
0                                       10        1  251.966782  USDC per ETH  
0  [0, 2, 4, 6, 20, 18, 8, 16, 14, 12, 10]       11  251.966782  USDC per ETH

And importantly we see the same price to the exact algo

The test now is to see the price get better as the request amount is lowered

In [14]:
results = FastSim.amm_sells('USDC',1500000000, execute=False)['trades']  # route_trade_by_target
assert(f"{results[results.uid=='3']['amt2'].values[0]:.0f}" == "5904869")
results

uid  id subid                                      note   aggr   exec  \
0  3.0   3     0                                  route #0  False  False   
0  3.1   3     1                                  route #2  False  False   
0  3.2   3     2                                  route #4  False  False   
0  3.3   3     3                                  route #6  False  False   
0  3.4   3     4                                 route #20  False  False   
0  3.5   3     5                                 route #18  False  False   
0    3   3     A  AMM sells 1500000000USDC buys 5904869ETH   True  False   

  limitfail          amt1  tkn1          amt2 tkn2     pair  \
0      None  2.548147e+08  USDC  9.934316e+05  ETH  ETHUSDC   
0      None  2.538271e+08  USDC  9.934543e+05  ETH  ETHUSDC   
0      None  2.528394e+08  USDC  9.934771e+05  ETH  ETHUSDC   
0      None  2.518518e+08  USDC  9.935000e+05  ETH  ETHUSDC   
0      None  2.508641e+08  USDC  9.935232e+05  ETH  ETHUSDC   
0      None  2.358029e+08  USDC  9.374828e+05  ETH  ETHUSDC   
0      None  1.500000e+09  USDC  5.904869e+06  ETH  ETHUSDC   

                routeix  nroutes       price        p_unit  
0                     0        1  254.027651  USDC per ETH  
0                     2        1  254.027651  USDC per ETH  
0                     4        1  254.027651  USDC per ETH  
0                     6        1  254.027651  USDC per ETH  
0                    20        1  254.027651  USDC per ETH  
0                    18        1  254.027651  USDC per ETH  
0  [0, 2, 4, 6, 20, 18]        6  254.027651  USDC per ETH

And indeed the rate gets better